felix/calc/encapsulation_resolver.go

```diff
@@ -0,0 +1,207 @@
+
+		k, ok := update.Key.(model.IPPoolKey)
+		if ok {
+			r.encapCalc.RemoveModelPool(k)
+		} else {
+			log.Infof("failed to convert %+v to model.IPPoolKey. Ignoring.", update.Key)
+		}
+	} else {
+		log.WithField("update", update).Debug("EncapsulationResolver: IPPool update")
+
+		pool, ok := update.Value.(*model.IPPool)
+		if ok {
+			r.encapCalc.UpdateModelPool(pool)
+		} else {
+			log.Infof("failed to convert %+v to *model.IPPool. Ignoring.", update.Value)
+		}
+	}
+
+	r.TriggerCalculation()
+
+	return
+}
+
+func (r *EncapsulationResolver) OnStatusUpdate(status api.SyncStatus) {
+	log.WithField("status", status).Debug("EncapsulationResolver: SyncStatus update")
+
+	if status == api.InSync {
+		r.inSync = true
+		r.TriggerCalculation()
+	}
+}
+
+func (r *EncapsulationResolver) TriggerCalculation() {
+	// Do nothing if EncapsulationResolver hasn't sync'ed all updates yet
+	if !r.inSync {
+		return
+	}
+
+	if r.config != nil {
+		newIPIPEnabled := r.encapCalc.IPIPEnabled()
+		newVXLANEnabled := r.encapCalc.VXLANEnabled()
+
+		if r.config.Encapsulation.IPIPEnabled != newIPIPEnabled || r.config.Encapsulation.VXLANEnabled != newVXLANEnabled {
+			log.WithFields(log.Fields{
+				"oldIPIPEnabled":  r.config.Encapsulation.IPIPEnabled,
+				"newIPIPEnabled":  newIPIPEnabled,
+				"oldVXLANEnabled": r.config.Encapsulation.VXLANEnabled,
+				"newVXLANEnabled": newVXLANEnabled,
+			}).Info("EncapsulationResolver: Encapsulation changed. Restart Felix.")
+			r.configChangedRestartCallback()
+		}
+	}
+}
+
+// EncapsulationCalculator is a helper struct to aid in calculating if IPIP and/or VXLAN
+// encapsulation should be enabled based on the existing IP Pools and their
+// configuration. It is used by EncapsulationResolver in this file, where it watches for
+// encapsulation changes to restart Felix, and by Run() in daemon.go, where it calculates
+// the encapsulation state that will be effectively used by Felix.
+type EncapsulationCalculator struct {
+	config     *config.Config
+	ipipPools  map[string]struct{}
+	vxlanPools map[string]struct{}
+}
+
+func NewEncapsulationCalculator(config *config.Config, ippoolKVList *model.KVPairList) *EncapsulationCalculator {
+	encapCalc := &EncapsulationCalculator{
+		config:     config,
+		ipipPools:  map[string]struct{}{},
+		vxlanPools: map[string]struct{}{},
+	}
+
+	if ippoolKVList != nil {
+		encapCalc.SetAPIPools(ippoolKVList)
+	}
+
+	return encapCalc
+}
+
+func (c *EncapsulationCalculator) UpdatePool(cidr string, ipipEnabled, vxlanEnabled bool) {
+	if ipipEnabled {
+		c.ipipPools[cidr] = struct{}{}
+	} else {
+		delete(c.ipipPools, cidr)
+	}
+
+	if vxlanEnabled {
+		c.vxlanPools[cidr] = struct{}{}
+	} else {
+		delete(c.vxlanPools, cidr)
+	}
+}
+
+func (c *EncapsulationCalculator) RemovePool(cidr string) {
+	delete(c.ipipPools, cidr)
+	delete(c.vxlanPools, cidr)
+}
+
+func (c *EncapsulationCalculator) UpdateAPIPool(pool *apiv3.IPPool) {
```

Comment:
I'm a bit surprised to see apiv3.IPPool show up here - I thought Felix was working entirely with `model.IPPool` objects. Where does this come from?

I feel like we should be able to reduce the number of functions attached to this but want to check.

Reply:
When listing the IP pools with `backendClient.List(ctx, model.ResourceListOptions{Kind: apiv3.KindIPPool}, "")` in daemon.go, I got KVPairs with apiv3.IPPool as the values. This (UpdateAPIPool) is only used by SetAPIPools(), which is used on initialization if the listed pools are passed (so it only happens in daemon.go)... Again, let me know if there is a way to list the pools and get a model type, I do agree that would make things much simpler. Thanks!